### DISCUSSION TITLE
If statement simplification

### DISCUSSION THREAD

**Discussion by DeLuxe-1337** [dotnet/csharplang#5621] [Language Ideas]
I was programming and thought of a little feature that would be useful in my opinion.

instead of doing

`if (node.Left.Kind() == SyntaxKind.StringLiteralExpression &&  node.Right.Kind() == SyntaxKind.StringLiteralExpression)`

you could do

`if(node.Left.Kind() && node.Right.Kind() == SyntaxKind.StringLiteralExpression)`

This would make lines shorter and it's efficient if you ask me!

**Comment by theunrepentantgeek** [comment]
Anyone used to one of the `{` braces `}` family of languages would expect `node.Left.Kind()` to have a bool value in the second example.

There are some fairly nasty interactions with existing valid code. For example, this would become ambiguous:

``` csharp
if (node.Left.Kind().IsLiteral && node.Right.Kind().IsLiteral == true)
```

(assuming `IsLiteral` is a **bool** property.)

You can already write this in a more compact form:

``` csharp
using static SyntaxKind;

if (node.Left.Kind() == StringLiteralExpression && node.Right.Kind() == StringLiteralExpression)
```

**Reply by DavidArno** [reply]
And if you want it even shorter:
```c#
if (node.Left.Kind() is var k && k == StringLiteralExpression && k == node.Right.Kind())
```
which demonstrates that shorter doesn't always mean better.

**Comment by huoyaoyuan** [comment]
The proposal makes `a == v && b == v` becomes `a && b == v`. If a and b is boolean type and v is `false`, the expression breaks.
Making the same expression works differently for bool and non-bool is definitely unacceptable.

**Comment by huoyaoyuan** [comment]
An interesting feature is to *match multiple values with one pattern*. Like `a is B and C`, can we make `a and b is C` work?

**Comment by macco3k** [comment]
I like the idea of making the language more terse, but without brackets that expression reads quite ambiguos. Maybe with a different syntax? E.g. `{a, b} &= value` for `all` behaviour and `{a, b} |= value` for `any` behaviour.

**Reply by FaustVX** [reply]
`&=` and `|=` already means something else, and can be confusing.
If I read `{a, b} &= value`, I will understand it as
```csharp
a = a & value;
b = b & value;
```

**Reply by macco3k** [reply]
> `&=` and `|=` already means something else, and can be confusing. If I read `{a, b} &= value`, I will understand it as
> 
> ```cs
> a = a & value;
> b = b & value;
> ```

Fair enough, always forget about bitwise operators. Still we already have `{}` initializers for lists and arrays, that's why I suggested curly brackets. You could also swap `&` and `=` so it's clear it's an equality operator
```
{a, b} ==& value
```
But really, it's quite tricky to find something that works. Ideally, I would just like to say something like `myValues.AnyEquals(value)`, which of course you can do by just writing your own `IEnumerable` extension method. I'd just like to be able to define arrays in a much terser way though (e.g. `{{a, b}}` or something instead of `new [] { }`).

**Reply by siegfriedpammer** [reply]
`==& expr` already means something else: https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAICuA7AzgQwDMBTdWdAYXQG9l16yMAjAexYBt0BZACgEtsAF3T4AlDToMpUAOzpsmdpwC8ygGQiA3JPoBfZLqA=

Besides, I'd prefer C# not to have too many obscure operators.

**Comment by Rekkonnect** [comment]
I think the best approach to this occurring problem would be a multi-way equality check syntax. For now, your best bet to mimic this behavior is a helper method that does this:
```csharp
public static bool AreEqual<T>(T a, T b, T c)
{
    // Assumes that b == c due to transitivity
    return a == b && a == c;
}
```

**Comment by NigelBess** [comment]
if(new[] {node.Left, node.Right}.All(n=>n.Kind() == StringLiteralExpression))

**Reply by Rekkonnect** [reply]
An array allocation and LINQ for a simple three-way equality check, that's just over the board.
